File: src/main/java/core/procedures/cons/ListTail.java
package core.procedures.cons;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Cons;
import core.scm.Type;

import java.util.List;

public final class ListTail extends AFn {

  public ListTail() {
    super(new FnArgsBuilder().min(2).max(2)
                             .mandatory(new Class[]{Object.class, Type.ExactNonNegativeInteger.class}).build());
  }

  @Override
  public String getName() {
    return "list-tail";
  }

  @Override
  public Object apply2(Object arg1, Object arg2) {
    Long p = ((Number)arg2).longValue();
    if (p == 0) {
      return arg1;
    }
    if (!(arg1 instanceof List)) {
      throw new WrongTypeException(getName(), "List", arg1);
    }
    List list = (List) arg1;
    if (p >= list.size() + 1) {
      throw new IndexOutOfBoundsException(String.format("%s: value out of range: %s", getName(), p));
    }
    /* Cons cell */
    if ((list instanceof Cons) && !((Cons)list).isList()) {
      if (p == 1) {
        return ((Cons)list).cdr();
      } else {
        throw new IndexOutOfBoundsException(String.format("%s: value out of range: %s", getName(), p));
      }
    }
    return list.subList(p.intValue(), list.size());
  }
}

File: src/main/java/core/procedures/cons/ListTail.kt
package core.procedures.cons

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.scm.Cons
import core.scm.Type

class ListTail : AFn<Any?, Any?>(name = "list-tail", minArgs = 2, maxArgs = 2,
                     mandatoryArgsTypes = arrayOf(Any::class.java, Type.ExactNonNegativeInteger::class.java)) {

    private val cdr = Cdr()

    override operator fun invoke(arg1: Any?, arg2: Any?): Any? {
        val p = (arg2 as Number).toLong()
        if (p == 0L) {
            return arg1
        }
        val list = arg1 as? List<*> ?: throw WrongTypeException(name, "List", arg1)
        if (p >= list.size + 1) {
            throw IndexOutOfBoundsException("$name: value out of range: $p")
        }
        /* Cons cell */
        if (!Cons.isProperList(list)) {
            if (p == 1L) {
                return cdr(list)
            } else {
                throw IndexOutOfBoundsException("$name: value out of range: $p")
            }
        }
        return list.subList(p.toInt(), list.size)
    }
}
